roachtest: activerecord failed [cached plan must not change result type]

roachtest.activerecord [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/20715249?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/20715249?buildTab=artifacts#/activerecord) on master @ [df1585efc36592c4f5ce8a73fd4c4cfe0aefb6f5](https://github.com/cockroachdb/cockroach/commits/df1585efc36592c4f5ce8a73fd4c4cfe0aefb6f5):


```
(orm_helpers.go:225).summarizeFailed: 
Tests run on Cockroach v26.1.0-alpha.00000000-dev-df1585efc36592c4f5ce8a73fd4c4cfe0aefb6f5
Tests run against activerecord 8.0.1
8558 Total Tests Run
8544 tests passed
14 tests failed
67 tests skipped
39 tests ignored
0 tests passed unexpectedly
13 tests failed unexpectedly
0 tests expected failed but skipped
0 tests expected failed but not run
---
--- FAIL: DelegatedTypeTest#test_delegated_type_predicates_with_custom_foreign_type (unexpected)
--- PASS: SerializedAttributeTest#test_unexpected_serialized_type (expected)
--- FAIL: DelegatedTypeTest#test_delegated_class_with_custom_foreign_type (unexpected)
--- FAIL: DelegatedTypeTest#test_scope_with_custom_foreign_type (unexpected)
--- FAIL: DelegatedTypeTest#test_scope (unexpected)
--- FAIL: DelegatedTypeTest#test_accessor (unexpected)
--- FAIL: DelegatedTypeTest#test_touch_account (unexpected)
--- FAIL: DelegatedTypeTest#test_association_id (unexpected)
--- FAIL: DelegatedTypeTest#test_delegated_class (unexpected)
--- FAIL: DelegatedTypeTest#test_delegated_type_predicates (unexpected)
--- FAIL: DelegatedTypeTest#test_builder_method (unexpected)
--- FAIL: DelegatedTypeTest#test_delegated_types (unexpected)
--- FAIL: DelegatedTypeTest#test_association_uuid (unexpected)
--- PASS: SerializedAttributeTestWithYamlSafeLoad#test_unexpected_serialized_type (expected)
--- FAIL: DelegatedTypeTest#test_delegated_type_name (unexpected)
For a full summary look at the activerecord artifacts 
An updated blocklist (activeRecordBlocklist) is available in the artifacts' activerecord log
test artifacts and logs in: /artifacts/activerecord/run_1
```

Parameters:
 - <code>arch=amd64</code>
 - <code>cloud=gce</code>
 - <code>coverageBuild=false</code>
 - <code>cpu=4</code>
 - <code>encrypted=false</code>
 - <code>fs=ext4</code>
 - <code>localSSD=true</code>
 - <code>metamorphicLeases=default</code>
 - <code>metamorphicWriteBuffering=true</code>
 - <code>runtimeAssertionsBuild=false</code>
 - <code>ssd=0</code>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)



See: [Grafana](https://go.crdb.dev/roachtest-grafana/teamcity-20715249/activerecord/1762758244102/1762760575081)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #156991 roachtest: activerecord failed [B-runtime-assertions-enabled C-test-failure O-roachtest O-robot T-sql-foundations branch-release-25.2 release-blocker]
- #156921 roachtest: activerecord failed [extra quoting] [C-test-failure O-roachtest O-robot P-2 T-sql-foundations branch-release-24.3.23-rc]
- #156912 roachtest: activerecord failed [C-test-failure O-roachtest O-robot P-2 T-sql-foundations branch-release-25.2.9-rc]
- #156441 roachtest: activerecord failed [extra quoting] [B-runtime-assertions-enabled C-test-failure O-roachtest O-robot P-2 T-sql-foundations branch-release-24.3 target-release-24.3.24]
- #155695 roachtest: activerecord failed [C-test-failure O-roachtest O-robot P-3 T-sql-foundations branch-release-25.3.3-rc]
- #152774 roachtest: activerecord failed [cached plan must not change result type] [B-runtime-assertions-enabled C-test-failure O-roachtest O-robot P-3 T-sql-foundations branch-release-25.3]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*activerecord.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-56376

**Note:** This build has runtime assertions enabled. If the same failure was hit in a run without assertions enabled, there should be a similar failure without this message. If there isn&#39;t one, then this failure is likely due to an assertion violation or (assertion) timeout.

roachtest.activerecord [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/20795811?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/20795811?buildTab=artifacts#/activerecord) on master @ [62084c233ab0355bfb5b87b172b363ad70333c9f](https://github.com/cockroachdb/cockroach/commits/62084c233ab0355bfb5b87b172b363ad70333c9f):

Failed with:

```
(orm_helpers.go:225).summarizeFailed: 
Tests run on Cockroach v26.1.0-alpha.1-dev-62084c233ab0355bfb5b87b172b363ad70333c9f
Tests run against activerecord 8.0.1
8556 Total Tests Run
8542 tests passed
14 tests failed
67 tests skipped
41 tests ignored
0 tests passed unexpectedly
13 tests failed unexpectedly
0 tests expected failed but skipped
0 tests expected failed but not run
---
--- FAIL: StrictLoadingTest#test_strict_loading_violation_on_polymorphic_relation (unexpected)
--- PASS: SerializedAttributeTest#test_unexpected_serialized_type (expected)
--- FAIL: SchemaTest#test_pk_and_sequence_for_with_schema_specified (unexpected)
--- FAIL: CockroachDB::FixturesResetPkSequenceTest#test_create_fixtures_resets_sequences_when_not_cached (unexpected)
--- FAIL: SchemaTest#test_set_pk_sequence (unexpected)
--- FAIL: CockroachDB::ConnectionAdapters::PostgreSQLAdapterTest#test_pk_and_sequence_for (unexpected)
--- FAIL: StrictLoadingTest#test_strict_loading_violation_logs_on_polymorphic_relation (unexpected)
--- FAIL: PostgresqlUUIDGenerationTest#test_pk_and_sequence_for_uuid_primary_key (unexpected)
--- FAIL: TestAutosaveAssociationOnAHasAndBelongsToManyAssociationWithAcceptsNestedAttributes#test_should_validation_the_associated_models_on_create (unexpected)
--- FAIL: CockroachDB::ConnectionAdapters::PostgreSQLAdapterTest#test_pk_and_sequence_for_with_non_standard_primary_key (unexpected)
--- FAIL: ActiveRecord::WhereTest#test_where_not_association_as_nand (unexpected)
--- PASS: SerializedAttributeTestWithYamlSafeLoad#test_unexpected_serialized_type (expected)
--- FAIL: PostgresqlRenameTableTest#test_renaming_a_table_with_uuid_primary_key_and_gen_random_uuid()_default_also_renames_the_primary_key_index (unexpected)
--- FAIL: PostgresqlRenameTableTest#test_renaming_a_table_with_uuid_primary_key_and_uuid_generate_v4()_default_also_renames_the_primary_key_index (unexpected)
--- FAIL: SchemaTest#test_reset_pk_sequence (unexpected)
For a full summary look at the activerecord artifacts 
An updated blocklist (activeRecordBlocklist) is available in the artifacts' activerecord log
test artifacts and logs in: /artifacts/activerecord/run_1
```
Cluster Node to Ip Mapping:
| Node | Public IP | Private IP |
| --- | --- | --- |
| teamcity-20795811-1764659597-14-n1cpu4-0001 | 34.148.245.17 | 10.142.3.79 |

Parameters:
 - <code>arch=amd64</code>
 - <code>cloud=gce</code>
 - <code>coverageBuild=false</code>
 - <code>cpu=4</code>
 - <code>encrypted=false</code>
 - <code>fs=ext4</code>
 - <code>localSSD=true</code>
 - <code>metamorphicLeases=default</code>
 - <code>runtimeAssertionsBuild=true</code>
 - <code>ssd=0</code>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)



See: [Grafana](https://go.crdb.dev/roachtest-grafana/teamcity-20795811/activerecord/1764659779152/1764662034612)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #157872 roachtest: activerecord failed [cached plan must not change result type] [B-runtime-assertions-enabled C-test-failure O-roachtest O-robot P-2 T-sql-foundations branch-release-25.4.0-rc]
- #157871 roachtest: activerecord failed [cached plan must not change result type] [C-test-failure O-roachtest O-robot P-2 T-sql-foundations branch-release-25.3.5-rc]
- #156991 roachtest: activerecord failed [B-runtime-assertions-enabled C-test-failure O-roachtest O-robot P-2 T-sql-foundations branch-release-25.2]
- #156921 roachtest: activerecord failed [extra quoting] [C-test-failure O-roachtest O-robot P-2 T-sql-foundations branch-release-24.3.23-rc]
- #152774 roachtest: activerecord failed [cached plan must not change result type] [B-runtime-assertions-enabled C-test-failure O-roachtest O-robot P-3 T-sql-foundations branch-release-25.3]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*activerecord.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

roachtest.activerecord [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/20803018?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/20803018?buildTab=artifacts#/activerecord) on master @ [1b46be83de2eb3ca447c3dc02f6b28c005526e6a](https://github.com/cockroachdb/cockroach/commits/1b46be83de2eb3ca447c3dc02f6b28c005526e6a):

Failed with:

```
(orm_helpers.go:225).summarizeFailed: 
Tests run on Cockroach v26.1.0-alpha.1-dev-1b46be83de2eb3ca447c3dc02f6b28c005526e6a
Tests run against activerecord 8.0.1
8556 Total Tests Run
8546 tests passed
10 tests failed
67 tests skipped
41 tests ignored
0 tests passed unexpectedly
9 tests failed unexpectedly
0 tests expected failed but skipped
0 tests expected failed but not run
---
--- PASS: SerializedAttributeTestWithYamlSafeLoad#test_unexpected_serialized_type (expected)
--- FAIL: PostgresqlUUIDGenerationTest#test_pk_and_sequence_for_uuid_primary_key (unexpected)
--- FAIL: PostgresqlRenameTableTest#test_renaming_a_table_with_uuid_primary_key_and_gen_random_uuid()_default_also_renames_the_primary_key_index (unexpected)
--- FAIL: SchemaTest#test_reset_pk_sequence (unexpected)
--- FAIL: SchemaTest#test_pk_and_sequence_for_with_schema_specified (unexpected)
--- FAIL: PostgresqlRenameTableTest#test_renaming_a_table_with_uuid_primary_key_and_uuid_generate_v4()_default_also_renames_the_primary_key_index (unexpected)
--- FAIL: SchemaTest#test_set_pk_sequence (unexpected)
--- PASS: SerializedAttributeTest#test_unexpected_serialized_type (expected)
--- FAIL: CockroachDB::ConnectionAdapters::PostgreSQLAdapterTest#test_pk_and_sequence_for (unexpected)
--- FAIL: CockroachDB::FixturesResetPkSequenceTest#test_create_fixtures_resets_sequences_when_not_cached (unexpected)
--- FAIL: CockroachDB::ConnectionAdapters::PostgreSQLAdapterTest#test_pk_and_sequence_for_with_non_standard_primary_key (unexpected)
For a full summary look at the activerecord artifacts 
An updated blocklist (activeRecordBlocklist) is available in the artifacts' activerecord log
test artifacts and logs in: /artifacts/activerecord/run_1
```
Cluster Node to Ip Mapping:
| Node | Public IP | Private IP |
| --- | --- | --- |
| teamcity-20803018-1764746186-08-n1cpu4-0001 | 35.196.74.176 | 10.142.0.216 |

Parameters:
 - <code>arch=amd64</code>
 - <code>cloud=gce</code>
 - <code>coverageBuild=false</code>
 - <code>cpu=4</code>
 - <code>encrypted=false</code>
 - <code>fs=ext4</code>
 - <code>localSSD=true</code>
 - <code>metamorphicLeases=default</code>
 - <code>runtimeAssertionsBuild=false</code>
 - <code>ssd=0</code>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)



See: [Grafana](https://go.crdb.dev/roachtest-grafana/teamcity-20803018/activerecord/1764746328148/1764749057744)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #158569 roachtest: activerecord failed [B-runtime-assertions-enabled C-test-failure O-roachtest O-robot P-2 T-sql-foundations branch-release-24.3]
- #157871 roachtest: activerecord failed [cached plan must not change result type] [C-test-failure O-roachtest O-robot P-2 T-sql-foundations branch-release-25.3.5-rc]
- #156991 roachtest: activerecord failed [B-runtime-assertions-enabled C-test-failure O-roachtest O-robot P-2 T-sql-foundations branch-release-25.2]
- #156921 roachtest: activerecord failed [extra quoting] [C-test-failure O-roachtest O-robot P-2 T-sql-foundations branch-release-24.3.23-rc]
- #152774 roachtest: activerecord failed [cached plan must not change result type] [B-runtime-assertions-enabled C-test-failure O-roachtest O-robot P-3 T-sql-foundations branch-release-25.3]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*activerecord.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/cmd/roachtest/tests/activerecord.go
+++ b/pkg/cmd/roachtest/tests/activerecord.go
@@ -28,8 +28,8 @@ var railsReleaseTagRegex = regexp.MustCompile(`^v(?P<major>\d+)\.(?P<minor>\d+)\
 
 // WARNING: DO NOT MODIFY the name of the below constant/variable without approval from the docs team.
 // This is used by docs automation to produce a list of supported versions for ORM's.
-var supportedRailsVersion = "8.0.1"
-var activerecordAdapterVersion = "v8.0.1"
+var supportedRailsVersion = "8.1.0"
+var activerecordAdapterVersion = "v8.1.0"
 
 // This test runs activerecord's full test suite against a single cockroach node.
 
@@ -173,7 +173,18 @@ func registerActiveRecord(r registry.Registry) {
 
 		t.Status("running activerecord test suite")
 
+		// Enable autocommit before DDL by default within the test suite.
 		result, err := c.RunWithDetailsSingleNode(ctx, t.L(), option.WithNodes(node),
+			`cd /mnt/data1/activerecord-cockroachdb-adapter/ && `+
+				`sudo sed -i 's/autocommit_before_ddl: false/autocommit_before_ddl: true/g' test/config.yml`,
+		)
+		// Fatal for a roachprod or transient error. A roachprod error is when result.Err==nil.
+		// Proceed for any other (command) errors
+		if err != nil && (result.Err == nil || rperrors.IsTransient(err)) {
+			t.Fatal(err)
+		}
+
+		result, err = c.RunWithDetailsSingleNode(ctx, t.L(), option.WithNodes(node),
 			fmt.Sprintf(
 				`cd /mnt/data1/activerecord-cockroachdb-adapter/ && `+
 					`sudo RAILS_VERSION=%s RUBYOPT="-W0" TESTOPTS="-v" bundle exec rake test`, supportedRailsVersion),

--- a/pkg/cmd/roachtest/tests/activerecord_blocklist.go
+++ b/pkg/cmd/roachtest/tests/activerecord_blocklist.go
@@ -36,6 +36,7 @@ var activeRecordIgnoreList = blocklist{
 	`AssociationCallbacksTest#test_has_many_callbacks_for_destroy_on_parent`:                                                                                   "flaky",
 	`BasicsTest#test_default_values_are_deeply_dupped`:                                                                                                         "flaky",
 	`CascadedEagerLoadingTest#test_eager_association_loading_with_cascaded_three_levels_by_ping_pong`:                                                          "flaky",
+	`CockroachDBReferentialIntegrityTest#test_should_reraise_invalid_foreign_key_exception_and_show_warning`:                                                   "affected by autocommit_before_ddl",
 	`CockroachDB::AdapterForeignKeyTest#test_foreign_key_violations_are_translated_to_specific_exception_with_validate_false`:                                  "flaky",
 	`CockroachDB::AdapterForeignKeyTest#test_foreign_key_violations_on_delete_are_translated_to_specific_exception`:                                            "flaky",
 	`CockroachDB::AdapterForeignKeyTest#test_foreign_key_violations_on_insert_are_translated_to_specific_exception`:                                            "flaky",

